File: Kotlin/ByTier/A/2109_A_It_s_Time_To_Duel.kt
// Problem: CF 2109 A - It's Time To Duel
// https://codeforces.com/contest/2109/problem/A

/*
 * Problem: A. It's Time To Duel
 * 
 * Purpose: Determine if at least one player must be lying based on their reported duel outcomes.
 * 
 * Algorithm:
 * - Model the tournament as a directed graph.
 * - Each player reports whether they won (1) or did not win (0) any duel.
 * - For each player, check constraints:
 *   - If a player reports 0: they must not have beaten anyone (no out edges), and cannot be beaten by anyone (no in edges).
 *   - If a player reports 1: they must have beaten at least one person or been beaten by someone.
 * 
 * Time Complexity: O(n)
 * Space Complexity: O(n)
 */

import java.util.*

fun main() {
    val x = readLine()!!.toInt()
    repeat(x) {
        val y = readLine()!!.toInt() // number of players
        val z = IntArray(y + 1) // player reports, index 1 to y
        val input = readLine()!!.split(" ").map { it.toInt() }
        for (i in 1..y) {
            z[i] = input[i - 1] // load reports into array starting at index 1
        }
        val w = y - 1 // number of duels (consecutive pairs)
        val q = Array(w + 1) { mutableListOf<Int>() } // adjacency list for players who could beat others
        val u = mutableListOf<Pair<Int, Boolean>>() // queue of candidates to process with direction info

        // Build graph and initial queue based on reports
        for (i in 1..y) {
            if (z[i] == 0) { // player did not win any duel
                if (i > 1) u.add(Pair(i - 1, true)) // can be beaten by left neighbor
                if (i < y) u.add(Pair(i, false)) // could beat right neighbor
            } else { // player won at least one duel
                when (i) {
                    1 -> u.add(Pair(1, true)) // first player can't be beaten but might have beaten next
                    y -> u.add(Pair(y - 1, false)) // last player can't beat anyone but might have been beaten
                    else -> q[i - 1].add(i) // add edge from i-1 to i (i might win against i-1)
                }
            }
        }

        val p = IntArray(w + 1) { -1 } // -1 means unvisited, 0 means not selected, 1 means selected
        val v: Deque<Pair<Int, Boolean>> = ArrayDeque() // BFS queue with direction

        for (e in u) v.add(e)

        var o = false // flag to store contradiction found
        while (v.isNotEmpty() && !o) {
            val (a, b) = v.removeFirst() // get next candidate node with direction

            if (p[a] != -1) { // if already processed
                if (p[a] != if (b) 1 else 0) { // check if conflict exists
                    o = true
                }
                continue
            }

            p[a] = if (b) 1 else 0 // mark as selected in direction b

            if (b) { // traverse forward
                for (t in q[a]) {
                    v.add(Pair(t, true))
                }
            }
        }

        println(if (o) "YES" else "NO") // print result: YES if contradiction found, NO otherwise
    }
}


// https://github.com/VaHiX/CodeForces/

File: CPP/ByTier/A/2109_A_It_s_Time_To_Duel.cpp
// Problem: CF 2109 A - It's Time To Duel
// https://codeforces.com/contest/2109/problem/A

/*
 * Problem: A. It's Time To Duel
 * Purpose: Determine if at least one player is lying about their duel results in a tournament.
 * 
 * Algorithm/Techniques:
 * - Simulation approach to check consistency of reports.
 * - For each test case, we simulate the duel outcomes and validate:
 *   1. No two consecutive players both report 0 (since at least one must win).
 *   2. The total number of reported wins (sum of array) is less than n (total number of duels).
 * 
 * Time Complexity: O(n) per test case, where n is the number of players.
 * Space Complexity: O(1), only using a few variables regardless of input size.
 */

#include <cstdio>
int main() {
  long t;
  scanf("%ld", &t);
  while (t--) {
    long n;
    scanf("%ld", &n);
    bool liar(false);                    // Flag to indicate if a liar is found
    long cnt(0), prev(1);                // cnt counts total wins, prev tracks previous player's report
    for (long p = 0; p < n; p++) {
      int x;
      scanf("%d", &x);
      if (!prev && !x) {                 // If both current and previous player report 0, contradiction
        liar = true;
      }
      prev = x;                          // Store current player's report for next iteration
      cnt += x;                          // Accumulate total wins reported
    }
    if (cnt >= n) {                      // If total reports >= number of players, impossible
      liar = true;
    }
    puts(liar ? "YES" : "NO");           // Output result
  }
}


// https://github.com/VaHiX/codeForces/